src/platform/w3dengine/client/w3ddebugicons.cpp

```diff
@@ -0,0 +1,263 @@
+#include "globaldata.h"
+#include "vertmaterial.h"
+#include <new>
+
+#ifdef BUILD_WITH_D3D8
+#include "dx8indexbuffer.h"
+#include "dx8vertexbuffer.h"
+#include "dx8wrapper.h"
+#endif
+
+DebugIcon *W3DDebugIcons::s_debugIcons = nullptr;
+int W3DDebugIcons::s_numDebugIcons = 0;
+int W3DDebugIcons::s_numUnknown = 0;
+
+W3DDebugIcons::W3DDebugIcons() : m_vertexMaterial(VertexMaterialClass::Get_Preset(VertexMaterialClass::PRELIT_DIFFUSE))
+{
+    Allocate_Icons_Array();
+}
+
+W3DDebugIcons::~W3DDebugIcons()
+{
+    Ref_Ptr_Release(m_vertexMaterial);
+
+    if (s_debugIcons != nullptr) {
+        delete[] s_debugIcons;
+        s_debugIcons = nullptr;
+    }
+
+    s_numDebugIcons = 0;
+    s_numUnknown = 0;
+}
+
+void W3DDebugIcons::Get_Obj_Space_Bounding_Sphere(SphereClass &sphere) const
+{
+    Vector3 pos(g_theWriteableGlobalData->m_waterExtentX, g_theWriteableGlobalData->m_waterExtentY, 500.0f);
+    sphere.Init(pos, pos.Length());
+}
+
+void W3DDebugIcons::Get_Obj_Space_Bounding_Box(AABoxClass &box) const
+{
+    Vector3 c(-2.0f * g_theWriteableGlobalData->m_waterExtentX, -2.0f * g_theWriteableGlobalData->m_waterExtentY, 0.0f);
+    Vector3 e(2.0f * g_theWriteableGlobalData->m_waterExtentX, 2.0f * g_theWriteableGlobalData->m_waterExtentY, 1000.0f);
+
+    box.Init(c, e);
+}
+
+RenderObjClass *W3DDebugIcons::Clone() const
+{
+    captainslog_warn("Cannot clone W3DDebugIcons!");
+    return nullptr;
+}
+
+void W3DDebugIcons::Render(RenderInfoClass &rinfo)
+{
+#if BUILD_WITH_D3D8
+    if (W3D::Are_Static_Sort_Lists_Enabled()) {
+        W3D::Add_To_Static_Sort_List(this, 1);
+    } else {
+        bool any_ended = false;
+
+        if (s_numDebugIcons) {
+            DX8Wrapper::Apply_Render_State_Changes();
+
+            DX8Wrapper::Set_Material(m_vertexMaterial);
+
+            DX8Wrapper::Set_Texture(0, nullptr);
+
+            DX8Wrapper::Apply_Render_State_Changes();
+
+            Matrix3D m(m_transform);
+            DX8Wrapper::Set_Transform(D3DTS_WORLD, m);
+
+            int num_icons = s_numDebugIcons;
+            if (s_numDebugIcons > 5000) {
+                num_icons = 5000;
+            }
+
+            int index = 0;
+
+            while (index < s_numDebugIcons) {
+                int index_count = 0;
+                int vertex_count = 0;
+
+                DynamicVBAccessClass vb_access(VertexBufferClass::BUFFER_TYPE_DYNAMIC_DX8, DX8_FVF_XYZNDUV2, 4 * num_icons);
+
+                DynamicIBAccessClass ib_access(IndexBufferClass::BUFFER_TYPE_DYNAMIC_DX8, 6 * num_icons);
+
+                {
+                    DynamicVBAccessClass::WriteLockClass lock_vb(&vb_access);
+                    VertexFormatXYZNDUV2 *verts = lock_vb.Get_Formatted_Vertex_Array();
+
+                    DynamicIBAccessClass::WriteLockClass lock_ib(&ib_access);
+
+                    if (verts == nullptr) {
+                        return;
+                    }
+
+                    unsigned short *indices = lock_ib.Get_Index_Array();
```

Comment:
`verts` is tested above, but `indices` is not. Is this proper?

Reply:
yes but added bugfixes on this